controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+)
+
+type coreDNSMigrator interface {
+	Migrate(currentVersion string, toVersion string, corefile string, deprecations bool) (string, error)
+}
+
+type CoreDNSMigrator struct{}
+
+func (c *CoreDNSMigrator) Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile string, deprecations bool) (string, error) {
+	return migration.Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile, deprecations)
+}
+
+type coreDNSInfo struct {
+	Corefile       string
+	Deployment     *appsv1.Deployment
+	CurrentVersion string
+}
+
+// UpdateCoreDNS updates the kubeadm configmap, coredns corefile and coredns
+// deployment.
+func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane) error {
+	info, err := w.getCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		return nil
+	}
+
+	// Validate the image tag.
+	if err := validateCoreDNSImageTag(info.CurrentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		return errors.Wrapf(err, "failed to validate CoreDNS")
+	}
+
+	// Get the DNS cluster configuration.
+	dns := &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+
+	// Perform the upgrade.
+	if err := w.updateCoreDNSImageInfoInKubeadmConfigMap(ctx, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSCorefile(ctx, info, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSDeployment(ctx, info, dns); err != nil {
+		return errors.Wrap(err, "unable to update coredns deployment")
+	}
+	return nil
+}
+
+// validateCoreDNSImageTag returns error if the versions don't meet requirements.
+// Some of the checks come from
+// https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
+func validateCoreDNSImageTag(fromVersion, toVersion string) error {
+	from, err := semver.Parse(fromVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+	}
+	to, err := semver.Parse(toVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+	}
+	if from.Compare(to) >= 0 {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	}
+
+	// check if the from version is even in the list of coredns versions
+	if _, ok := migration.Versions[fmt.Sprintf("%d.%d.%d", from.Major, from.Minor, from.Patch)]; !ok {
+		return fmt.Errorf("fromVersion %q is not a compatible coredns version", from.String())
+	}
+	return nil
+}
+
+// getCoreDNSInfo gets
+// - the current CoreDNS image version installed
+// - the current Corefile Configuration of CoreDNS
+// - the coredns deployment
+func (w *Workload) getCoreDNSInfo(ctx context.Context) (*coreDNSInfo, error) {
+	// Get the coredns configmap and corefile.
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	cm := &corev1.ConfigMap{}
+	if err := w.Client.Get(ctx, corednsKey, cm); err != nil {
```

Comment:
Should we attempt to fail a bit more gracefully in the event that Get returns a not found error?

Not that I actually expect someone to be running a cluster where CoreDNS isn't installed, but we handle kube-proxy that way and it gives the flexibility that one could replace that component if they know what they are doing.

Reply:
It also looks like we are attempting to fail more gracefully if coredns is not chosen in other places as well, doing so here would provide some consistency with those attempts.

Reply:
added a check in the caller so that if we get any "cause" that matches IsNotFound, it'll return right away without erroring